---
title: "Microarray Data - Rspondin KO/KI with or without infection in stomach corpus - DGE Reanalysis with LIMMA"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, results='hide', messages=FALSE}
rm(list=ls())
library(limma)
library(writexl)
library(pheatmap)
library(xtable)
library(reshape2)

load("../../Data/Processed/Corpus_Rspo_KI_and_KO_with_or_wo_Hp_micro_array_preprocessed_data.Rdata")

result_folder = "../../Results/Array1/"
if (!file.exists(result_folder)) dir.create(result_folder, recursive=T)
```

# Introduction

This is data from stomach corpus samples of various mouse experiments (Rspo KO/KI, Infection or both vs controls) hybridized as dual channel libraries to Agilent 014568 micro arrays. 

# MDS

```{r}
MA.avg <- avereps(MA, ID=MA$genes$ProbeName)
exp_matrix = MA.avg$M * matrix(rep(ifelse(ed$dye_swap, -1, 1),each=nrow(MA.avg$M)),nrow(MA.avg$M),ncol(MA.avg$M))
colnames(exp_matrix) = ed[colnames(exp_matrix), "array_label"]
ed_exp_mat = ed
rownames(ed_exp_mat) = ed$array_label
```


```{r, fig.width=8, fig.height=8}
################################################################################
## MDS on normalized data 
cp = palette(rainbow(11))
data_inp = t(exp_matrix)

d <- dist(data_inp) # euclidean distances between the rows
fit <- cmdscale(d,eig=TRUE, k=2) # k is the number of dim

# plot solution
x <- fit$points[,1]
y <- fit$points[,2]
cc = cp[as.numeric(factor(ed_exp_mat[rownames(data_inp),]$Condition))]
plot(x, y, xlab="Coordinate 1", ylab="Coordinate 2", main="Metric MDS, all samples", type="n", ylim=c(min(y)-50, max(y)+50), xlim=c(min(x)-50, max(x)+200))
points(x, y, col=cc)
#text(x,y,labels=ed[rownames(data_inp),]$sample_ID, col=cp[as.numeric(factor(ed[rownames(data_inp),]$Tissue.Status))])
text(x,y,labels=ed_exp_mat[rownames(data_inp),]$Short, col=cc, cex=.8, pos=4)
################################################################################
```

# Diffential gene expression

Standard dual-channel analysis of differential expression data using LIMMA is applied for comparisons. 

We do the following comparisons: 

  - Rspo3 KI vs. WT in mice after 14d without infection
  - Rspo3 KO vs. WT after 2m infection with H.pylori
  
```{r}
all_results = list()

controls = MA$genes$ControlType!=0
MA.avg <- avereps(MA[!controls,], ID=MA[!controls,]$genes$ProbeName)

```

```{r}
#######################################################################
sel_samples = rownames(subset(ed, Experiment == "Rspo KI vs WT 14d non infected"))
ed_tmp = ed[sel_samples,]
design = ifelse(ed_tmp$Cy5_condition=="Rspo3 KI", 1, -1)
fit <- lmFit(MA.avg[,sel_samples], design)
fit <- eBayes(fit)
res = topTable(fit, adjust = "BH",number=nrow(fit))

# ratio_tab = sweep(MA.avg[,sel_samples]$M, 2, design, "*")
# colnames(ratio_tab) = ed_tmp$array_label
# res = merge(res, ratio_tab, all.x=T, by.x="ProbeName", by.y=0, sort=F)
all_results[["Rspo_KI_vs_WT_14d_non_infected"]] = res

#######################################################################
sel_samples = rownames(subset(ed, Experiment == "Rspo  KO vs WT 2m infected"))
ed_tmp = ed[sel_samples,]
design = ifelse(ed_tmp$Cy5_condition=="Rspo3 KO", 1, -1)
fit <- lmFit(MA.avg[,sel_samples], design)
fit <- eBayes(fit)
res = topTable(fit, adjust = "BH",number=nrow(fit))

# ratio_tab = sweep(MA.avg[,sel_samples]$M, 2, design, "*")
# colnames(ratio_tab) = ed_tmp$array_label
# res = merge(res, ratio_tab, all.x=T, by.x="ProbeName", by.y=0, sort=F)
all_results[["Rspo_KO_vs_WT_2m_infected"]] = res
```

## Volcano plots

Due to the small replicate number no genes in any of the comparisons reached a significance level of FDR < 5%. We here therefore show unadjusted (raw) p-values noting that **a large part or all of them might be false positive calls**.

```{r, volcano, fig.width=8, fig.height=8}
all_target_conditions = names(all_results)
par(mfrow=c(2,2), mar=c(8,4,4,2))

for (tc in all_target_conditions) {
  r = all_results[[tc]]
  plot(r$logFC, -log10(r$P.Value),xlab="log2 Fold Change",ylab="-log10(raw p-val)", ylim=c(0,max(2,max(-log10(r$P.Value),na.rm=T))))
  title(main=tc, sub=paste("(",nrow(subset(r, P.Value < 0.05))," signif. DE genes)",sep="") )
  abline(h=-log10(0.05),col="red")
  abline(v=c(-1,1))
}
```

```{r, DE_combined, results="hide"}
###############################################################################################################
# Write Result Files
###############################################################################################################

all_DE_results_tmp = list()
for (tc in all_target_conditions) {
  tmp = all_results[[tc]]
  tmp$condition = tc
  all_DE_results_tmp[[tc]] = tmp
}
all_DE_results_ts = do.call(rbind, all_DE_results_tmp)
all_DE_results_ts$DE_class = ifelse(all_DE_results_ts$adj.P.Val>0.01, "not sign", ifelse(all_DE_results_ts$logFC>0,ifelse(all_DE_results_ts$logFC>1,"Up","Up_weak"), ifelse(all_DE_results_ts$logFC < -1,"Down", "Down_weak")))
agg_fun = function(x) ifelse("Down" %in% x, "Down",ifelse("Up" %in% x, "Up","not_significant."))
all_DE_results_sw = dcast(all_DE_results_ts, GeneSymbol ~ condition, value.var="DE_class", fun.aggregate=agg_fun)
```


```{r, echo=FALSE}
R.avg <- avereps(RG$R, ID=RG$genes$ProbeName)
G.avg <- avereps(RG$G, ID=RG$genes$ProbeName)
colnames(R.avg) = paste(ed[colnames(R.avg),]$Cy5,"_Cy5_label",sep="")
colnames(G.avg) = paste(ed[colnames(G.avg),]$Cy3,"_Cy3_label",sep="")

intensity_matrix = cbind(R.avg, G.avg)
norm_intensity_matrix = normalizeBetweenArrays(intensity_matrix, method="quantile")
```


```{r, write_tabs, echo=FALSE}


output_file_prefix = paste(result_folder,"Differential_expression_results_", sep="/")
selected_cols = c("ProbeName", "GeneSymbol", "GeneName","EntrezID","logFC","AveExpr","t","P.Value","adj.P.Val","GencodeM12MappingStatus","GeneSymbol_Gencode","Transcript_Type" )
for (tc in all_target_conditions) {
  write.table(all_results[[tc]][,selected_cols], file=paste(output_file_prefix, tc, ".txt", sep="" ), row.names=F , sep="\t", dec=".")
}

short_names = list()
for (i in names(all_results)) {
  ii = gsub(" ","_", gsub(",","_",i))
  short_names[[i]] = ii
}

DGE_Excel_filename = file.path(result_folder,"DiffExpression.xlsx")
write_xlsx(all_results, path = DGE_Excel_filename)


##############################################
# Normalized expression values per sample
anno = unique(RG$genes[,!colnames(RG$genes) %in% c("Row","Col","Start","index")])
rownames(anno) = anno$ProbeName
norm_exp_mat = data.frame(ProbeName = rownames(norm_intensity_matrix), log2(norm_intensity_matrix), row.names=NULL, stringsAsFactors = F)
norm_exp_mat$GeneSymbol = anno[as.character(norm_exp_mat$ProbeName), "GeneSymbol"]
filename = paste(result_folder,"Normalized_expression_data.txt",sep="/")
write.table(norm_exp_mat, file=filename,sep="\t",row.names=F)

```


```{r}
output_file = paste(result_folder,"DGE_analysis_image.Rdata", sep="/")
save(all_results, RG, MA, ed, file=output_file)

filename = paste(result_folder,"DGE_results.Rdata",sep="/")
exp_design = ed
save(all_results, exp_design, norm_intensity_matrix, file=filename)
```


# Session Info
```{r}
sessionInfo()
```