---
title: "Microarray Data - Rspondin KI after 2m with infection, 14d KI without infection and Rspo KO after 2m infection in stomach corpus - GSEA analysis on DGE results"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, warning=FALSE, message=FALSE}
rm(list=ls())

library(reshape2)
library(pheatmap)
library(ggplot2)
library(knitr)
library(scales)
library(fgsea)
library(data.table)
library(readxl)
library(writexl)
library(BiocParallel)

result_folder = "../../Results/GSEA/"

set.seed(12345)
```

# Overview

This is data from a stomach corpus samples of mouse experiments (Rspo KI/KO with/without infection vs controls) hybridized as dual channel libraries to Agilent 074809 (SurePrint G3 Mouse GE v2 8x60K) micro arrays or Agilent 014568 micro arrays. 

For analysis, a preranked analysis using the R package fGSEA is used that should give similar results to preranked analysis in standard GSEA. 

# MSigDB gene sets

Since MSigDB gene sets use human gene symbols to map genes to pathways we will translate mouse symbols to homologous human symbols using HomologeneDB from NCBI (build 68). 


```{r}
r1_env = new.env()
r2_env = new.env()

load("../../Results/Array1/DGE_results.Rdata", envir=r1_env)
load("../../Results/Array2/DGE_results.Rdata", envir=r2_env)

all_results = list()

tmp = get("all_results", r1_env)
all_results[["Rspo_KO_vs_WT_2m_infected"]] = tmp[["Rspo_KO_vs_WT_2m_infected"]]
all_results[["Rspo_KI_vs_WT_14d_non_infected"]] = tmp[[ "Rspo_KI_vs_WT_14d_non_infected"]]

tmp = get("all_results", r2_env)

all_results[["Rspo_KI_vs_WT_2m_infected"]] = tmp[["Rspo_KI_vs_WT_2m_infected"]]

```

```{r}
gsea_gmt_folder = "../../Data/External/GeneSets/MSigDB_v7.1/"
gsea_gmt_files = c("h.all.v7.1.symbols.gmt","c2.all.v7.1.symbols.gmt","c3.all.v7.1.symbols.gmt","c5.bp.v7.1.symbols.gmt","c6.all.v7.1.symbols.gmt","c7.all.v7.1.symbols.gmt")
names(gsea_gmt_files) = c("HALLMARK","PATHWAYS","MOTIF","GO_BP","ONCOGENIC","IMMUNOLOGIC")
gset_collections_human = list()
for (n in names(gsea_gmt_files)) {
  gset_collections_human[[n]] = gmtPathways(paste(gsea_gmt_folder, gsea_gmt_files[[n]], sep="/") )
}

```

We use gene sets from MSigDB.

```{r}
gsea_gmt_files
```


# Other gene sets 

## Stem cell/beta catenin signatures

  - Munoz et al Small Intestinal Stem Cell Signature (PMID: 22692129, DOI: 10.1038/emboj.2012.166, Supp Table S3 )
  - DSS Sca+ vs. Non-DSS (Yui et al. Cell Stem Cell 2017, p.35; derived from published microarray data) 
  - SPEM signature (Nozaki et al. Gastroenterology 2008, p.511; supp tables 2/3)
  

Note that the Parietal signature from Ramsey et al contains a large fraction of interferon signaling associated genes. Since there is only a single replicate in this data set it is not clear if this is an artefact or real biological differences. 

  
```{r}
known_genes = unique(all_results[[1]]$GeneSymbol)


########################################################################################
custom_gene_sets_mouse = list()

spem_sig = read.table("../../Data/External/Nozaki_et_al_2008/SPEM_signature.txt", sep="\t", header=T, stringsAsFactors = F)
custom_gene_sets_mouse[["Nozaki_SPEM_up"]] = subset(spem_sig, direction == "Up" & !is.na(GeneSymbol))$GeneSymbol
custom_gene_sets_mouse[["Nozaki_SPEM_down"]] = subset(spem_sig, direction == "Down" & !is.na(GeneSymbol))$GeneSymbol


tmp = read.table("../../Data/External/GeneSets/Supp_Table_S3_GE_combined_SC_Signature.txt", sep="\t", header=T, stringsAsFactors = F)
custom_gene_sets_mouse[["Lgr5_SC_Signature_Munoz"]] = unique(tmp$GeneSymbol)

tmp = read.table("../../Data/External/Yui_et_al_CSC_2018/Colitis_vs_control/DGE_Colitis_vs_control.txt", sep="\t", header=T, stringsAsFactors = F)

custom_gene_sets_mouse[["Yui_DSS_colitis_vs_control_up"]] = subset(tmp, logFC > 1 & adj.P.Val < 0.05)$SYMBOL
custom_gene_sets_mouse[["Yui_DSS_colitis_vs_control_down"]] = subset(tmp, logFC < -1 & adj.P.Val < 0.05)$SYMBOL

gset_collections_mouse = list(CUSTOM_MOUSE=custom_gene_sets_mouse)

```

## fGSEA analysis 

```{r}
homologene_human_mouse = fread("../../Data/External/homologene_9606_10090.txt")
all_entrez_ids = data.frame(EntrezID=unique(all_results[[1]]$EntrezID))
all_symbols = data.frame(GeneSymbol=unique(all_results[[1]]$GeneSymbol))
a1 = merge(all_symbols, homologene_human_mouse[,c("entrez_1","symbol_1","symbol_2"), with=F], by.x="GeneSymbol", by.y="symbol_2", all.x=T, sort=F)

mouse_to_human=a1
mouse_to_human = subset(mouse_to_human, !is.na(GeneSymbol) & !is.na(symbol_1) & !symbol_1 == "")
rownames(mouse_to_human) = mouse_to_human$GeneSymbol
```


### Human and Mouse gene sets

For analyzing the human MsigDB gene sets we restrict the DGE data to those probesets that have a homologous gene in mouse and human. For those probesets we select the one with the highest t-score and rank the resulting list which is subsequently used for fGSEA analysis. 

For mouse gene sets (i.e. those derived from mouse experiments) we use the full set of genes in the DGE results after collapsing t-scores by gene and ranking. 

```{r}
output_folder = "../../Results/GSEA"
GSEA_datafile = file.path(output_folder,"fGSEA_Results.Rdata")
if (!file.exists(output_folder)) dir.create(output_folder, recursive=T)

rerun_GSEA = TRUE

if(rerun_GSEA) {
  all_gsea_results = list()
  n_perm = 5000
  n_proc = 8
  bpparam = MulticoreParam(worker=n_proc, manager.hostname="127.0.0.1")
  for (n in names(all_results)) {
    print(paste("Running FGSEA for comparison",n))
    gsea_result_tables = list()
    tmp = subset(all_results[[n]], !is.na(GeneSymbol)) # those are mostly control probes
    
    tmp2 = merge(tmp, mouse_to_human, by="GeneSymbol")
    tmp2 = subset(tmp2, !is.na(symbol_1))
    ranks = unlist(tapply(tmp2$t, tmp2$symbol_1, function(x) x[which(abs(x)==max(abs(x)))] ))
  
    for (gs in names(gset_collections_human)) {
      print(gs)
      fgseaRes <- fgsea(gset_collections_human[[gs]], ranks, minSize=15, maxSize=2000, nperm=n_perm, BPPARAM=bpparam)
      fgseaRes$GeneSetCollection = gs
      gsea_result_tables[[gs]] = fgseaRes
    }
    
    results_human = do.call(rbind, gsea_result_tables)
    
    gsea_result_tables_mouse = list()
    ranks_mouse = unlist(tapply(tmp$t, tmp$GeneSymbol, function(x) x[which(abs(x)==max(abs(x)))] ))
    
    for (gs in names(gset_collections_mouse)) {
      print(gs)
      fgseaRes <- fgsea(gset_collections_mouse[[gs]], ranks_mouse, minSize=15, maxSize=2000, nperm=n_perm, BPPARAM=bpparam)
      fgseaRes$GeneSetCollection = gs
      gsea_result_tables_mouse[[gs]] = fgseaRes
    }
  
    results_mouse = do.call(rbind, gsea_result_tables_mouse)
    results_all = rbind(results_human, results_mouse)
    results_all$FDR_global = p.adjust(results_all$pval, "BH")
    
    all_gsea_results[[n]] = results_all
  }
  save(all_gsea_results, file=GSEA_datafile)
  save(gset_collections_human, gset_collections_mouse, file=file.path(output_folder, "Geneset_collections.Rdata"))
} else {
  load(GSEA_datafile)
}
```

Result files are written to folder `r result_folder`

We export all gene sets with less than 10% FDR after global adjustment of p-values per sample. 

```{r}
rearrange_gsea <- function(x, FDR_thr = 0.1) {
    x$direction= ifelse(x$NES > 0, "Up","Down")
    x = x[order(x$NES, decreasing=T),]
    x$leadingEdge = unlist(lapply(x$leadingEdge, function(x) paste(x, collapse=",")))
    x = subset(x, FDR_global < FDR_thr)
    col_order = c("pathway","GeneSetCollection","direction","pval","padj","ES","NES","nMoreExtreme","size","leadingEdge", "FDR_global")
    x[, col_order, with=F]
}
```


```{r}
FDR_threshold = 0.1
if(rerun_GSEA) {
  output_file_prefix = paste(result_folder,"fGSEA_Results_", sep="/")
  for (n in names(all_gsea_results)) {
    filename = paste(output_file_prefix, n, ".txt", sep="" )
    tmp = rearrange_gsea(all_gsea_results[[n]], FDR_threshold)
    write.table(tmp, file= filename, row.names=F , sep="\t", dec=".", quote=F, na="")
  }
  
  tmp = lapply(all_gsea_results, rearrange_gsea, FDR_threshold)
  GSEA_Excel_filename = file.path(result_folder,"fGSEA_Results.xlsx")
  write_xlsx(tmp, path=GSEA_Excel_filename)
  
}
```

```{r}
if(rerun_GSEA) {
  output_file_prefix = paste(result_folder,"fGSEA_Results_all_comparisons", sep="/")
  agr = list()
  for (n in names(all_gsea_results)) {
    tmp = all_gsea_results[[n]]
    tmp$direction= ifelse(tmp$NES > 0, "Up","Down")
    tmp = tmp[order(tmp$NES, decreasing=T),]
    tmp$leadingEdge = unlist(lapply(tmp$leadingEdge, function(x) paste(x, collapse=",")))
    tmp = subset(tmp, FDR_global < FDR_threshold )
    #col_order = c("pathway","GeneSetCollection","direction","pval","padj","ES","NES","nMoreExtreme","size","leadingEdge","FDR_global")
    tmp$Comparison = n
    agr[[n]] = tmp
  }
  
  agr_combined = do.call(rbind, agr)
  
  filename = paste(output_file_prefix, ".txt", sep="" )
  write.table(agr_combined, file= filename, row.names=F , sep="\t", dec=".", quote=F, na="")
  
  filename = file.path(result_folder,"fGSEA_Results_matrix_NES_FDR_lt_0.05.txt")
  gsea_mat = dcast(agr_combined, GeneSetCollection + pathway ~ Comparison, value.var = "NES")
  write.table(gsea_mat, file = filename, row.names=F, sep="\t", dec=".", quote = F, na="")
  
}
```


```{r, fig.width=12, fig.height=8}
gsea_summary = list()
for (n in names(all_gsea_results)) {
  tmp = subset(all_gsea_results[[n]], FDR_global < FDR_threshold)
  if (nrow(tmp)==0) {
    gsea_summary[[n]] = data.frame(Collection=c("Empty","Empty"), Direction=c("Up", "Down"), Freq=rep(0,2), Comparison = n)  
    next
  }
  r = as.data.frame(table(tmp$GeneSetCollection, ifelse(tmp$ES<0, "Down","Up")))
  colnames(r) = c("Collection","Direction","Freq")
  r$Comparison = n
  gsea_summary[[n]] = r
}

gsea_summary_df = do.call(rbind, gsea_summary)

ggplot(gsea_summary_df, aes(x=Comparison, y = Freq, fill=Collection)) + geom_bar(stat="identity") + theme_grey(base_size = 14) + theme(axis.text.x = element_text(angle = 60, hjust = 1)) + ggtitle("Number of significant (FDR<10%) gene sets per comparison")

ggplot(gsea_summary_df, aes(x=Comparison, y = Freq, fill=Collection)) + geom_bar(stat="identity") + theme_grey(base_size = 14) + theme(axis.text.x = element_text(angle = 60, hjust = 1)) + ggtitle("Number of significant (FDR<10%) gene sets per comparison") + facet_wrap(~Direction)
```




```{r}
sessionInfo()
```


